stan/math/opencl/kernels/normal_id_glm_lpdf.hpp

```diff
@@ -0,0 +1,161 @@
+        __global double* y_minus_mu_over_sigma_squared_sum,
+        __global double* sigma_derivative, __global double* log_sigma_sum,
+        const __global double* y, const __global double* x,
+        const __global double* alpha, const __global double* beta,
+        const __global double* sigma_glob, const int N, const int M,
+        const int is_alpha_vector, const int is_sigma_vector,
+        const int need_mu_derivative, const int need_mu_derivative_sum,
+        const int need_sigma_derivative, const int need_log_sigma_sum) {
+      const int gid = get_global_id(0);
+      const int lid = get_local_id(0);
+      const int lsize = get_local_size(0);
+      const int wgid = get_group_id(0);
+
+      __local double res_loc[LOCAL_SIZE_];
+
+      double y_minus_mu_over_sigma_squared = 0;
+      double log_sigma = 0;
+      double mu_derivative = 0;
+      if (gid < N) {
+        double y_minus_mu_over_sigma = 0;
+        for (int i = 0, j = 0; i < M; i++, j += N) {
+          y_minus_mu_over_sigma += x[j + gid] * beta[i];
+        }
+        double sigma = sigma_glob[gid * is_sigma_vector];
+        double inv_sigma = 1 / sigma;
+        y_minus_mu_over_sigma
+            = (y[gid] - y_minus_mu_over_sigma - alpha[gid * is_alpha_vector])
+              * inv_sigma;
+        mu_derivative = inv_sigma * y_minus_mu_over_sigma;
+        if (need_mu_derivative) {
+          mu_derivative_glob[gid] = mu_derivative;
+        }
+        y_minus_mu_over_sigma_squared
+            = y_minus_mu_over_sigma * y_minus_mu_over_sigma;
+        if (need_sigma_derivative) {
+          sigma_derivative[gid]
+              = (y_minus_mu_over_sigma_squared - 1) * inv_sigma;
+        }
+        if (need_log_sigma_sum) {
+          log_sigma = log(sigma);
+        }
+      }
+      res_loc[lid] = y_minus_mu_over_sigma_squared;
+      barrier(CLK_LOCAL_MEM_FENCE);
+      for (int step = lsize / REDUCTION_STEP_SIZE; step > 0;
+           step /= REDUCTION_STEP_SIZE) {
+        if (lid < step) {
+          for (int i = 1; i < REDUCTION_STEP_SIZE; i++) {
+            res_loc[lid] += res_loc[lid + step * i];
+          }
+        }
+        barrier(CLK_LOCAL_MEM_FENCE);
+      }
+      if (lid == 0) {
+        y_minus_mu_over_sigma_squared_sum[wgid] = res_loc[0];
+      }
+
+      if (need_mu_derivative_sum) {
+        barrier(CLK_LOCAL_MEM_FENCE);
+        res_loc[lid] = mu_derivative;
+        barrier(CLK_LOCAL_MEM_FENCE);
+        for (int step = lsize / REDUCTION_STEP_SIZE; step > 0;
+             step /= REDUCTION_STEP_SIZE) {
+          if (lid < step) {
+            for (int i = 1; i < REDUCTION_STEP_SIZE; i++) {
+              res_loc[lid] += res_loc[lid + step * i];
+            }
+          }
+          barrier(CLK_LOCAL_MEM_FENCE);
+        }
+        if (lid == 0) {
+          mu_derivative_sum[wgid] = res_loc[0];
+        }
+      }
+
+      if (need_log_sigma_sum) {
+        barrier(CLK_LOCAL_MEM_FENCE);
+        res_loc[lid] = log_sigma;
```

Comment:
log_sigma is only used here why is it calculated so far above?

Reply:
It is calculated within `if (gid < N) {`. I did not want to duplicate this if.